flock_unix.go

```diff
@@ -10,18 +10,33 @@ import (
 	"syscall"
 )
 
-// Lock is a blocking call to try and take the file lock. It will wait until it
-// is able to obtain the exclusive file lock. It's recommended that TryLock() be
-// used over this function. This function may block the ability to query the
-// current Locked() status due to a RW-mutex lock.
+// Lock is a blocking call to try and take an exclusive file lock. It will wait
+// until it is able to obtain the exclusive file lock. It's recommended that
+// TryLock() be used over this function. This function may block the ability to
+// query the current Locked() or RLocked() status due to a RW-mutex lock.
 //
-// If we are already locked, this function short-circuits and returns immediately
-// assuming it can take the mutex lock.
+// If we are already exclusive-locked, this function short-circuits and returns
+// immediately assuming it can take the mutex lock.
 func (f *Flock) Lock() error {
+	return f.lock(&f.l, syscall.LOCK_EX)
```

Comment:
@theory I've started to think about this approach of using the shared lock function, and I think in its current form it has a bit of a race condition issue because we have only one `Unlock()` method but two different ways to unlock the mutex. 

It seems that in some cases, the OS is smart enough to know whether or not a shared lock can become an exclusive lock and it may give the exclusive lock after giving us the shared lock. This is why the test was working on Linux, but stalling on Windows until we called `.Unlock()` before locking again.

The case that concerns me is when multiple things are using this, and maybe using different lock types. If one goroutine takes a shared lock, and something else takes an exclusive lock, it seems like the lock could transparently become an exclusive lock and then the shared locker could accidentally drop the exclusive lock without knowing it. Because `lock()` takes one of the locked boolean values, it never knows if we have an exclusive or shared lock and can't prevent the possibility of hitting this failure mode.

I'm not totally sure if this is possible, but I'm thinking it could potentially cause some unexpected behaviors / data corruption. What do you think?